copycat/runtime/src/main/java/org/apache/kafka/copycat/runtime/distributed/DistributedHerder.java

```diff
@@ -17,304 +17,590 @@
+                syncConfigsTimeoutMs = config.getInt(DistributedHerderConfig.WORKER_SYNC_TIMEOUT_MS_CONFIG);
+                backoffOnFailure = true;
+            } else if (configState.offset() < assignment.offset()) {
+                log.warn("Join group completed, but assignment failed and we lagging. Reading to end of config and retrying.");
+                needsReadToEnd = true;
+            } else {
+                log.warn("Join group completed, but assignment failed. We were up to date, so just retrying.");
+            }
+        } else {
+            if (configState.offset() < assignment.offset()) {
+                log.warn("Catching up to assignment's config offset.");
+                needsReadToEnd = true;
+            }
         }
-    }
 
-    private void createConnectorTasks(ConnectorState state) {
-        String taskClassName = state.connector.taskClass().getName();
-
-        log.info("Creating tasks for connector {} of type {}", state.name, taskClassName);
-
-        List<Properties> taskConfigs = state.connector.taskConfigs(state.maxTasks);
-
-        // Generate the final configs, including framework provided settings
-        Map<ConnectorTaskId, Properties> taskProps = new HashMap<>();
-        for (int i = 0; i < taskConfigs.size(); i++) {
-            ConnectorTaskId taskId = new ConnectorTaskId(state.name, i);
-            Properties config = taskConfigs.get(i);
-            // TODO: This probably shouldn't be in the Herder. It's nice to have Copycat ensure the list of topics
-            // is automatically provided to tasks since it is required by the framework, but this
-            String subscriptionTopics = Utils.join(state.inputTopics, ",");
-            if (state.connector instanceof SinkConnector) {
-                // Make sure we don't modify the original since the connector may reuse it internally
-                Properties configForSink = new Properties();
-                configForSink.putAll(config);
-                configForSink.setProperty(SinkTask.TOPICS_CONFIG, subscriptionTopics);
-                config = configForSink;
+        if (needsReadToEnd) {
+            log.info("Current config state offset {} is behind group assignment {}, reading to end of config log", configState.offset(), assignment.offset());
+            try {
+                configStorage.readToEnd().get(syncConfigsTimeoutMs, TimeUnit.MILLISECONDS);
+                configState = configStorage.snapshot();
+                log.info("Finished reading to end of log and updated config snapshot, new config log offset: {}", configState.offset());
+            } catch (TimeoutException e) {
+                log.warn("Didn't reach end of config log quickly enough", e);
+                // TODO: With explicit leave group support, it would be good to explicitly leave the group *before* this
+                // backoff since it'll be longer than the session timeout
+                if (backoffOnFailure)
```

Comment:
Actually, it's even just if (isLeader()) because if you are leader and need to catch up, that must mean you failed to make an assignment. Good catch, all those flags were getting confusing.